File: 2019-2020_programmers/_16_/src/java/MinDistBetweenNode.java

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Vector;
import java.util.stream.Collectors;

public class MinDistBetweenNode {
    public static void main(String[] args) {
        TreeNode root = new TreeNode(4);
        root.left = new TreeNode(2);
        root.left.left = new TreeNode(1);
        root.left.right = new TreeNode(3);
        root.right = new TreeNode(6);
        /* 예제 1/
        root = [4, 2, 6, 1, 3, null, null]
         */
        /* 예제 2/
        root = [90, 69, null, 49, 89, null, 52, null, null, null, null]
        */
        /*
        System.out.println(traverse(new Vector<Integer>(), root));
        */
    }
    public static int traverse(Vector<Integer> nums, TreeNode node) {
        /*
        System.out.println("---> stack opened @t node: " + node.val);
        */
        int min = Integer.MAX_VALUE;
        int diff = Integer.MAX_VALUE;
        if (node.left != null) {
            min = Math.min(traverse(nums, node.left), min);
            /* log/
            System.out.println("\t* left sub: " + min);
            */
        }
        /* log/
        System.out.println(nums);
        System.out.println("\tcur node: " + node.val);
        */
        if (nums.size() > 0) {
            diff = Math.abs(nums.lastElement() - node.val);
            /* log/
            System.out.println("\tdiff: " + diff);
            */
        }
        nums.add(node.val);
        if (node.right != null) {
            min = Math.min(traverse(nums, node.right), min);
            /*log/
            System.out.println("\t* right sub: " + min);
             */
        }
        min = Math.min(diff, min);
        /* log/
        System.out.println("<--- stack closed @t node: " + node.val + ", min :" + min + "\n");
         */
        return min;
    }

}

File: 2019-2020_programmers/_16_/src/kotlin/MinDistBetweenNode.kt
import java.util.Vector;
import kotlin.math.abs
import kotlin.math.min

fun main(args: Array<String>) {
   val root = TreeNode(4);
   root.left = TreeNode(2);
   root.left!!.left = TreeNode(1);
   root.left!!.right = TreeNode(3);
   root.right = TreeNode(6);
    /* 예제 1/
    root = [4, 2, 6, 1, 3, null, null]
     */
    /* 예제 2/
    root = [90, 69, null, 49, 89, null, 52, null, null, null, null]
    */
    /*
    */
   println(MinDistBetweenNode().traverse(Vector<Int>(), root));
}

public class MinDistBetweenNode {
   fun traverse(nums: Vector<Int>, node: TreeNode): Int {
       println("---> stack opened @t node: " + node.val)
        /*
        */
       var min = Integer.MAX_VALUE
       var diff = Integer.MAX_VALUE
       if (node.left != null) {
           /* java/ min = Math.min(traverse(nums, node.left!!), min) */
           min = traverse(nums, node.left!!).coerceAtMost(min)
           /* log/
           */
           println("\t* left sub: $min")
       }
       println(nums)
       println("\tcur node: " + node.value)
       /* log/
       */
       if (nums.size > 0) {
           /* java/  diff = Math.abs(nums.lastElement() - node.value) */
           diff = abs(nums.lastElement() - node.value)
           /* log/
           */
           println("\tdiff: $diff")
       }
       nums.add(node.value)
       node.right.let {  }
       if (node.right != null) {
           /* java/ min = Math.min(traverse(nums, node.right!!), min) */
           min = min(traverse(nums, node.right!!), min);
           /*log/
            */
           println("\t* right sub: $min")
       }
       /* java/ min = Math.min(diff, min) */
       min = min(diff, min);
       /* log/
        */
       println( "<--- stack closed @t node: $node.value, min : $min\n" );
       return min;
   }
}
